raft-tech/TANF-app
# Migrate to GitHub Native Dependabot

### Summary of Changes
Currently our Snyk configuration is configured on the `carltonsmith` Snyk organization which will no longer work after @carltonsmith leaves the project. Additionally, these Snyk PRs require us to manage an unnecessary `requirements.txt` file in addition to our Pipfile and python dependency update PRs are not complete when they get opened since they don't update anything in our actual dependencies. [Example Snyk Python Dependency PR](https://github.com/raft-tech/TANF-app/pull/903)

Additionally, we currently use the Dependabot Preview app which is being deprecated in favor of a [GitHub Native Dependabot](https://github.blog/2021-04-29-goodbye-dependabot-preview-hello-dependabot/) which has more features and is configured via a YAML file committed to the repo.

Rather than setting up Snyk on a new organization and in order to get ahead on the impending Dependabot migration this PR proposes an update which provides the necessary YAML config to enable the new GitHub Native version of Dependabot. The new version is capable of providing both [dependency version upgrades](https://docs.github.com/en/code-security/supply-chain-security/about-dependabot-version-updates) as well as [security upgrades in response to vulnerabilities](https://docs.github.com/en/code-security/supply-chain-security/about-dependabot-security-updates), which covers the functionality provided by both Snyk and Dependabot Preview.

NOTE: We may want to explicitly disable automated PR updates on the HHS repo if it is not already, otherwise once this file gets merged in we will get Dependabot PRs to both repos, which would conflict with our git flow as they would be opened on divergent branches. This can be disabled [in the security analysis settings](https://github.com/raft-tech/TANF-app/settings/security_analysis) as shown below:
![disable-dependabot-security-updates](https://user-images.githubusercontent.com/22626085/118340020-8b744f80-b4e8-11eb-8bb1-eb851f074627.png)

### How to Test
Unfortunately due to the way Dependabot works we can't fully test this until it gets merged. On merge to raft-tdp-main a scan should begin immediately and the status checks should update shortly thereafter [in the Dependency Graph](https://docs.github.com/en/code-security/supply-chain-security/enabling-and-disabling-version-updates#checking-the-status-of-version-updates)

We can however, run the proposed YAML file against the Dependabot provided validator [here](https://dependabot.com/docs/config-file/validator/) as shown below:
![Screenshot from 2021-05-14 18-06-14](https://user-images.githubusercontent.com/22626085/118336165-223c0e80-b4df-11eb-8224-de7e0745876f.png)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] Project dependencies can be automatically kept up to date.
  - [x] Security vulnerabilities in project dependencies can be automatically resolved.
  - [x] All relevant documentation is updated to reflect this change.

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A changes relate only to dependency and vulnerability management

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

N/A changes relate only to dependency and vulnerability management

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A no UX changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

N/A changes relate only to dependency and vulnerability management

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
